// zocket/core/client.js
var cmd = require('../command/cmd');
var Connector = require('../network/connector');
var request = require('../command/request');

function Client(host, port) {
  this.connector = new Connector(host, port);
  var self = this;
  this.connector.eventEmitter.on('processPacket', function(packet){
    self.dispatch(packet);
  });
}

Client.prototype = {
  start: function() {
    this.connector.connect();
    this.connector.send(request.requestHandShake());
  },

  send: function(request) {
    this.connector.send(request);
  },

  dispatch: function(response) {
    switch (response.cmdId) {
      case cmd.HAND_SHAKE:
        console.log('Hand shake');
        this.login();
        break;
      case cmd.LOGIN:
        console.log('Login successfully ^_^');
        break;
      default:
        this.dispatchExtended(response);
    }
  },

  login: function() {
    console.log('[-] Login with default method (using username only). You should write your own login method!');
    this.send(request.requestLogin());
  },

  setDispatch: function(method) {
    this.dispatchExtended = method;
  }
}

module.exports = Client
